language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -184,228 +126,471 @@ pub struct RootAccessPath {
-        match self {
-            Local::Ref(r) => Some(Local::Value(r.read_reference())),
-            Local::GlobalRef(gr) => Some(Local::Value(gr.read_reference())),
+impl From<Value> for Option<GlobalRef> {
+    fn from(value: Value) -> Option<GlobalRef> {
+        match value.0 {
+            ValueImpl::GlobalRef(reference) => Some(reference),
             _ => None,
         }
     }
+}
 
-    pub fn mutate_reference(self, v: MutVal) {
-        match self {
-            Local::Ref(r) => r.mutate_reference(v),
-            Local::GlobalRef(r) => r.mutate_reference(v),
-            _ => (),
+impl MutVal {
+    fn new(v: Value) -> Self {
+        MutVal(Rc::new(RefCell::new(v.0)))
+    }
+
+    fn peek(&self) -> Ref<ValueImpl> {
+        self.0.borrow()
+    }
+
+    fn into_value(self) -> VMResult<Value> {
+        match Rc::try_unwrap(self.0) {
+            Ok(cell) => Ok(Value::new(cell.into_inner())),
+            Err(_) => Err(VMStatus::new(StatusCode::LOCAL_REFERENCE_ERROR)),
         }
     }
 
-    pub fn value(self) -> Option<MutVal> {
-        match self {
-            Local::Value(v) => Some(v),
-            _ => None,
+    fn copy_value(&self) -> VMResult<Value> {
+        self.peek().copy_value()
+    }
+
+    fn size(&self) -> AbstractMemorySize<GasCarrier> {
+        self.peek().size()
+    }
+
+    fn borrow_field(&self, field_offset: usize) -> VMResult<Value> {
+        self.peek().borrow_field(field_offset)
+    }
+
+    fn write_value(self, value: Value) {
+        self.0.replace(value.0);
+    }
+
+    #[allow(non_snake_case)]
+    #[doc(hidden)]
+    fn to_type_FOR_TESTING(&self) -> Type {
+        self.peek().to_type_FOR_TESTING()
+    }
+
+    fn equals(&self, v2: &MutVal) -> VMResult<bool> {
+        self.peek().equals(&v2.peek())
+    }
+}
+
+impl Struct {
+    /// Creates a struct from a vector of `Value`s.
+    pub fn new(values: Vec<Value>) -> Self {
+        let mut fields = vec![];
+        for value in values {
+            fields.push(MutVal::new(value));
         }
+        Struct(fields)
     }
 
-    pub fn value_as<T>(self) -> Option<T>
-    where
-        Option<T>: From<MutVal>,
-    {
-        self.value().and_then(std::convert::Into::into)
+    /// Called by `Unpack` to fetch all fields out of the struct being unpacked.
+    pub fn get_field_value(&self, field_offset: usize) -> VMResult<Value> {
+        if let Some(field_ref) = self.0.get(field_offset) {
+            field_ref.copy_value()
+        } else {
+            Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR))
+        }
     }
 
-    pub fn size(&self) -> AbstractMemorySize<GasCarrier> {
-        match self {
-            Local::Ref(v) => v.size(),
-            Local::GlobalRef(v) => v.size(),
-            Local::Value(v) => v.size(),
-            Local::Invalid => *CONST_SIZE,
+    // Public because of gas synthesis - review.
+    pub fn get_field_reference(&self, field_offset: usize) -> VMResult<Value> {
```

Comment:
Is it the same as `BorrowField`? Plus if it is only needed by gas synthesis we can probably put a compilation flag 
`#[cfg(feature = "instruction_synthesis")]`

Reply:
I want to review the gas cost story and we have something planned with @tzakian to talk about it so a comment seemed just as fine for me now. Thoughts?